providers/common/der/der_writer.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2020 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include "internal/cryptlib.h"
+#include "crypto/bn.h"
+#include "prov/der.h"
+
+/*
+ * All functions return the same:
+ * == 0 : error
+ * >  0 : total length so far
+ */
+
+static size_t int_der_alloc(unsigned char **pp, const unsigned char *start,
```

Comment:
So I gonna be honest and say I find this a bit frustrating.
@paulidale can you review please.
@davidmakepeace  got a lot of resistance when he implemented a DER writer that didnt use PACKET to place inside the fips boundary.
So here we are again with yet another DER writer (YADW).

Reply:
I tend to agree with @slontis, we chose to use WPACKET for this kind of thing, we should stick to that choice (or redo the previous changes).

C really is an ugly language for byte processing :(

Reply:
I recall suggesting the approach I'm using here back then. But also, there was the belief that all we're going to need is a DSA signature writer and not a library... I never liked the WPACKET choice, but there were too many other voices...

Reply:
I can rewrite the DSA signature writer, sure.

Reply:
Honestly, I don't care which paradigm we use for building DER data into the FIPS boundary.  What I care about is that there is only one paradigm used.

I seem to remember there was only one voice strongly in favour of WPACKET, and the rest mostly didn't object too much.

Reply:
... that being said, I'd rather do that in a separate PR, as that's not *essential* for this PR, nor #11422, which I'm targetting with this.

Reply:
There was a discussion, and the decision was WPACKET. To have another PR that undoes that decision isn't a good look. :)